aws-android-sdk-pinpoint/src/main/java/com/amazonaws/mobileconnectors/pinpoint/targeting/notification/EventSourceType.java

```diff
@@ -0,0 +1,194 @@
+    }
+
+    private static final EventSourceAttributeParser CAMPAIGN_PARSER = new EventSourceAttributeParser() {
+        @Override
+        public Map<String, String> getEventSourceAttributes(Bundle bundle) {
+            if(bundle == null) {
+                return null;
+            }
+            final Map<String, String> campaignAttributes = new HashMap<>();
+            String campaignKeyPrefix = CAMPAIGN.getEventSourceKeyPrefix();
+
+            for(String key : bundle.keySet()) {
+                if (key.startsWith(campaignKeyPrefix)) {
+                    //If it matches the prefix, include it in the attributes
+                    //after stripping out the prefix
+                    campaignAttributes.put(key.replace(campaignKeyPrefix,""), bundle.getString(key));
+                }
+            }
+            return campaignAttributes.size() > 0 ? campaignAttributes : null;
+        }
+    };
+
+    private static final EventSourceAttributeParser JOURNEY_PARSER = new EventSourceAttributeParser() {
+        @Override
+        public Map<String, String> getEventSourceAttributes(Bundle bundle) {
+            String pinpointPayload = bundle.getString(PINPOINT_ATTRIBUTE_KEY);
+            if(pinpointPayload == null) {
+                return null;
+            }
+            Map<String, String> result = null;
+            try {
+                JsonElement pinpointJson = new JsonParser().parse(pinpointPayload);
+                JsonElement journeyJson = pinpointJson.getAsJsonObject().get(JOURNEY_ATTRIBUTE_KEY);
+                if(journeyJson == null) {
+                    return null;
+                }
+                result = new HashMap<>();
+                for(Map.Entry entry : journeyJson.getAsJsonObject().entrySet()) {
+                    result.put((String) entry.getKey(), ((JsonPrimitive)entry.getValue()).getAsString());
+                }
+            } catch(JsonSyntaxException ex) {
+                Log.w(TAG, "Exception attempting to parse pinpoint JSON payload.", ex);
+                Log.v(TAG, "Payload: "+pinpointPayload);
+                return null;
+            }
+
+            return result;
+        }
+    };
+
+    public static EventSourceType CAMPAIGN = new EventSourceType("campaign",
+            "_campaign",
+            CAMPAIGN_ID_ATTRIBUTE_KEY,
+            CAMPAIGN_ACTIVITY_ID_ATTRIBUTE_KEY,
+            CAMPAIGN_PUSH_KEY_PREFIX,
+            CAMPAIGN_PARSER);
+    public static EventSourceType JOURNEY = new EventSourceType("journey",
+            "_journey",
+            JOURNEY_ID_ATTRIBUTE_KEY,
+            JOURNEY_ACTIVITY_ID_ATTRIBUTE_KEY,
+            null,
+            JOURNEY_PARSER);
+
+    public static EventSourceType getEventSourceType(Bundle bundle) {
+        if(bundle == null) {
+            return null;
+        }
+        if(bundle.containsKey(CAMPAIGN.getEventSourceKeyPrefix()+CAMPAIGN.getEventSourceIdAttributeKey())) {
```

Comment:
`WhitespaceAround` here ... suggest to cache the keycompound to cleanup the `if`,

```
String campaignKey = // form the key compound .../
if (bundle.contains(campaignKey)) {
  // It's a campaign! Wooo!
} else {
  // distinctly not a campaign.
}
```
Same for the next repeated code block int he helper method below.

Reply:
Done. I also modified this to return an UNKNOWN event source. Returning journey on else felt a bit misleading.